### ISSUE TITLE
Slow query when searching a value in a big table and sorting is remembered

### ISSUE THREAD

**Issue by liviuconcioiu** [phpmyadmin/phpmyadmin#18942]
### Describe the bug

I know this is supposed to do, but when sorting is remembered (`Main panel > Browse mode > Remember tables's sorting` checked) and you search for a value in a big table, `ORDER BY` is added to each query and you can even wait for minutes for a query to execute, even if one row will be returned.

Hopefully there's a way for this to avoid `ORDER BY` when table sorting is being remembered?

### To Reproduce

1. Check Main panel > Browse mode > Remember tables's sorting
2. Search a value in a big table.
3. See the query being slow.

### Expected behavior

Improve the query execution time.

### Screenshots

With `ORDER BY`:

![2b](https://github.com/phpmyadmin/phpmyadmin/assets/25424343/f391f429-0d81-4918-82fb-1a9b1506dc99)
![1a](https://github.com/phpmyadmin/phpmyadmin/assets/25424343/635bfe6f-b6ca-4959-9ba7-439e2bbe7296)
![2c](https://github.com/phpmyadmin/phpmyadmin/assets/25424343/979607ab-10f7-45d9-b944-7484a142540e)

Without `ORDER BY`:

![2a](https://github.com/phpmyadmin/phpmyadmin/assets/25424343/fa947c3c-e9d9-402f-80fe-3c0fc921c882)

### Server configuration

- phpMyAdmin version: 5.2.2-dev, 6.0.0-dev

**Comment by kamil-tekiela** [comment]
I am not sure exactly how to reproduce this. I don't understand why `ORDER BY` would slow down the query. Is `id` a primary key?

**Comment by liviuconcioiu** [comment]
> I am not sure exactly how to reproduce this. I don't understand why `ORDER BY` would slow down the query. Is `id` a primary key?

`id` isn't a primary key in this case. Is just another bigint column with no index on it (there's no need for index because is used to store a bigint value that isn't used for lookup). I've just renamed the field so I can post the issue. 

I've edited the images to reflect a better picture. To reproduce this is very simple:

Using the attached table, run those two queries (with different `test_id` value to avoid query caching) and compare the time:

```SQL
SELECT * FROM `tbl_test` WHERE `test_id` = 10004978673
SELECT * FROM `tbl_test` WHERE `test_id` = 10004978673 ORDER BY `id` ASC
```

[tbl_test.zip](https://github.com/phpmyadmin/phpmyadmin/files/14073759/tbl_test.zip)

**Comment by kamil-tekiela** [comment]
Are you able to import this file using PMA? I keep getting "Incorrect format parameter"

**Comment by liviuconcioiu** [comment]
> Are you able to import this file using PMA? I keep getting "Incorrect format parameter"

Yeah, it works. Import has been successfully finished, 3111 queries executed. (tbl_test.sql).

You can create a sample table, with 10 million rows:

```SQL
CREATE TABLE `tbl_test` (
  `id` int(10) UNSIGNED NOT NULL,
  `test_id` bigint(20) UNSIGNED DEFAULT NULL
) ENGINE=MyISAM DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci ROW_FORMAT=FIXED;

INSERT INTO `tbl_test` (`id`, `test_id`) VALUES
(1, 1),
(2, 2),
(3, 3),
(4, 4),
...
```

**Comment by kamil-tekiela** [comment]
I still can't reproduce this. Is this MariaDB or MySQL?

**Comment by liviuconcioiu** [comment]
> I still can't reproduce this. Is this MariaDB or MySQL?

MariaDB, but it doesn't matter, because technically isn't a bug. Because `Remember tables's sorting` is checked, if you sort a table either `ASC` or `DESC` and then search for a value in a column that doesn't have an index, when searching, it will add `ORDER BY` to the query.

Not sure what it's possible to do here. Probably other people want the result sorted, but at the same time, if the result returns only one row, it takes ~20 seconds to get it.

Here's a video:

https://github.com/phpmyadmin/phpmyadmin/assets/25424343/89c33769-05bf-4a5b-9f83-2c1eba8ce13f

**Comment by kamil-tekiela** [comment]
Exactly. I was considering if we can disable this, but I doubt it. The search feature should be able to employ `ORDER BY`. 

No matter what I search for, I get the result back in less than 0.5s.

**Comment by liviuconcioiu** [comment]
> No matter what I search for, I get the result back in less than 0.5s.

Have you tried to enter different numbers, to avoid query caching? The thing is I get the same thing on localhost and on a few servers. All of them are using MyISAM. I don't know if this is the issue.

**Comment by kamil-tekiela** [comment]
I have tried the same steps again without changing anything and now it takes 12s. 
I tried with InnoDB too and it doesn't seem to suffer from this. 

It's a bit odd that adding `ORDER BY` affects the performance so much. Especially without `LIMIT`. The engine should first find the matching rows, which it can do quickly, and then sort the results. With a `LIMIT` it might take a slightly different path, but should still be optimized. I don't know why it sometimes take a lot longer to find the results. 

If I use PMA to run the two queries, I get less than 1s for first and 12s for the second one. But if I do the same in MariaDB console I get 0.2s for each of them. If I add `LIMIT 1` and use `DESC`, it takes 12s. Does PMA silently add the LIMIT?

I don't believe in query cache and I doubt that anything like this affects the performance here. 

I am not sure what to think about this issue. It's probably not related to PMA and it looks like there is nothing we can do. But I may be very wrong.

**Comment by liviuconcioiu** [comment]
> If I use PMA to run the two queries, I get less than 1s for first and 12s for the second one. But if I do the same in MariaDB console I get 0.2s for each of them. If I add `LIMIT 1` and use `DESC`, it takes 12s. Does PMA silently add the LIMIT?

Yeah, phpMyAdmin adds the `LIMIT 0, 25` to the query. Results are similar if `LIMIT` is added to the console.

phpMyAdmin:

```
Showing rows 0 - 0 (1 total, Query took 1.1061 seconds.)
SELECT * FROM `tbl_test` WHERE `test_id` = 7890123;

Showing rows 0 - 0 (1 total, Query took 10.8260 seconds.)
SELECT * FROM `tbl_test` WHERE `test_id` = 7890124 ORDER BY `id` ASC;
```

MariaDB:

```
MariaDB [test]> SELECT * FROM `tbl_test` WHERE `test_id` = 7890123 LIMIT 0, 25;
+---------+---------+
| id      | test_id |
+---------+---------+
| 7890128 | 7890128 |
+---------+---------+
1 row in set (0.670 sec)

MariaDB [test]> SELECT * FROM `tbl_test` WHERE `test_id` = 7890124 ORDER BY `id` ASC LIMIT 0, 25;
+---------+---------+
| id      | test_id |
+---------+---------+
| 7890125 | 7890125 |
+---------+---------+
1 row in set (10.493 sec)
```

> I don't believe in query cache and I doubt that anything like this affects the performance here.

Query caching is very effective, at least on my side. Take a look:

https://github.com/phpmyadmin/phpmyadmin/assets/25424343/0b4aeb00-0ad3-41a1-8b55-7615c2c90361

> I am not sure what to think about this issue. It's probably not related to PMA and it looks like there is nothing we can do. But I may be very wrong.

The only thing that can be really changed is the documentation, to mention about this issue with MyISAM and if you search in a column with no index. There's no word about search, only browse.

https://github.com/phpmyadmin/phpmyadmin/blob/0061af6fbf48fc25606d7a0346b31fad524c8935/docs/config.rst#L3196-L3201

Thanks for taking a look into this!

**Comment by williamdes** [comment]
Hi all,

What is left to be done on this one ?

**Comment by liviuconcioiu** [comment]
I ended up with `$cfg['RememberSorting'] = false;`.

**Comment by williamdes** [comment]
> I ended up with `$cfg['RememberSorting'] = false;`.

Okay, is there anything to do one this issue ?
Can it be closed ?

**Comment by liviuconcioiu** [comment]
> > I ended up with `$cfg['RememberSorting'] = false;`.
> 
> Okay, is there anything to do one this issue ? Can it be closed ?

No. Yes, it can be closed.
